flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/InputGateWithMetrics.java

```diff
@@ -0,0 +1,119 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.taskmanager;
+
+import org.apache.flink.metrics.Counter;
+import org.apache.flink.runtime.event.TaskEvent;
+import org.apache.flink.runtime.io.network.partition.consumer.BufferOrEvent;
+import org.apache.flink.runtime.io.network.partition.consumer.InputGate;
+import org.apache.flink.runtime.io.network.partition.consumer.InputGateListener;
+import org.apache.flink.runtime.metrics.groups.TaskIOMetricGroup;
+
+import java.io.IOException;
+import java.util.Optional;
+
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+/**
+ * The InputGateWithMetrics is wrapped in task with {@link InputGate} provided by
+ * shuffle service. It is mainly used for increasing general input metrics from
+ * {@link TaskIOMetricGroup}.
+ */
+public class InputGateWithMetrics implements InputGate, InputGateListener {
```

Comment:
Not sure why we need to implement `InputGateListener` as well.
Could we not just forward `registerListener` to `inputGate`?
Who is going to call `notifyInputGateNonEmpty` in this class?

Reply:
I also not considered this issue before, only found it when existing ITCase fails.

The `UnionInputGate` could only see `InputGateWithMetrics` and then put them into internal `inputGateToIndexOffsetMap`. When the `SingleInputGate` receives data from network, it would notify the listener(`UnionInputGate`) and then the `SingleInputGate` would be added into `UnionInputGate#inputGatesWithData`. When `UnionInputGate#getNextBufferOrEvent`, the `SingleInputGate` is fetched from `inputGatesWithData`, but it can not be found from `inputGateToIndexOffsetMap` which would cause NPE issue.

Reply:
I see, thanks for explanation!